src/Server/ArrowFlightHandler.cpp

```diff
@@ -0,0 +1,348 @@
+#include "ArrowFlightHandler.h"
+
+#include <memory>
+#include <arrow/result.h>
+#include <arrow/flight/api.h>
+#include <arrow/flight/types.h>
+#include <arrow/status.h>
+#include <arrow/memory_pool.h>
+#include <arrow/compute/api.h>
+#include <arrow/ipc/api.h>
+#include <arrow/table.h>
+#include <arrow/util/macros.h>
+
+#include <Common/logger_useful.h>
+#include <Common/Exception.h>
+#include <Core/Block.h>
+#include <Processors/Executors/PullingPipelineExecutor.h>
+#include <Processors/Formats/Impl/ArrowBufferedStreams.h>
+#include <Processors/Formats/Impl/ArrowColumnToCHColumn.h>
+#include <Interpreters/executeQuery.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/Session.h>
+#include <Server/IServer.h>
+#include <IO/ReadBufferFromString.h>
+#include <IO/WriteBufferFromString.h>
+#include <Interpreters/DatabaseCatalog.h>
+#include <Storages/IStorage.h>
+#include <Processors/Sinks/SinkToStorage.h>
+#include <Processors/ISource.h>
+#include <QueryPipeline/QueryPipelineBuilder.h>
+#include <QueryPipeline/Pipe.h>
+#include <QueryPipeline/QueryPipeline.h>
+#include <Processors/Executors/PipelineExecutor.h>
+#include <QueryPipeline/Chain.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Processors/Sources/SourceFromSingleChunk.h>
+#include <Processors/Executors/CompletedPipelineExecutor.h>
+#include <QueryPipeline/QueryPipeline.h>
+#include <QueryPipeline/Pipe.h>
+#include <Common/SettingsChanges.h>
+#include <Interpreters/InterpreterInsertQuery.h>
+#include <Parsers/ASTInsertQuery.h>
+#include <Parsers/ASTExpressionList.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Access/Credentials.h>
+#include <Processors/Formats/Impl/CHColumnToArrowColumn.h>
+
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int UNKNOWN_EXCEPTION;
+    extern const int CANNOT_CONVERT_TYPE;
+}
+
+ArrowFlightHandler::ArrowFlightHandler(IServer & server_, const Poco::Net::SocketAddress & address_to_listen_)
+    : server(server_)
+    , log(getLogger("ArrowFlightHandler"))
+    , address_to_listen(address_to_listen_)
+{
+    auto parse_location_status = arrow::flight::Location::ForGrpcTcp(address_to_listen_.host().toString(), address_to_listen_.port());
+    if (!parse_location_status.ok()) {
+        throw Exception(ErrorCodes::UNKNOWN_EXCEPTION,
+                        "Invalid address {} for Arrow Flight Server: {}",
+                        address_to_listen_.toString(),
+                        parse_location_status->ToString());
+    }
+    location = std::move(parse_location_status).ValueOrDie();
+}
+
+void ArrowFlightHandler::start() {
+    setThreadName("ArrowFlight");
+
+    session = std::make_unique<Session>(server.context(), ClientInfo::Interface::ARROW_FLIGHT);
+    session->authenticate("default", "", address_to_listen);
```

Comment:
So we also authenticate as user `default` with no password.

Reply:
added auth and tls support.

Authorization may be not working without tls for secure reasons in some clients